lnwire/dyn_propose.go

```diff
@@ -0,0 +1,249 @@
+			DPChannelReserveSatoshis, dp.ChannelReserve))
+	}
+	if dp.HtlcMinimum != nil {
+		tlvRecords = append(tlvRecords, tlv.MakePrimitiveRecord(
+			DPHtlcMinimumMsat, dp.HtlcMinimum))
+	}
+	if dp.CsvDelay != nil {
+		tlvRecords = append(tlvRecords, tlv.MakePrimitiveRecord(
+			DPToSelfDelay, dp.CsvDelay))
+	}
+	if dp.MaxAcceptedHTLCs != nil {
+		tlvRecords = append(tlvRecords, tlv.MakePrimitiveRecord(
+			DPMaxAcceptedHtlcs, dp.MaxAcceptedHTLCs))
+	}
+	if dp.FundingKey != nil {
+		tlvRecords = append(tlvRecords, tlv.MakePrimitiveRecord(
+			DPFundingPubkey, dp.FundingKey))
+	}
+	if dp.ChannelType != nil {
+		tlvRecords = append(tlvRecords, tlv.MakeDynamicRecord(
+			DPChannelType, dp.ChannelType,
+			dp.ChannelType.featureBitLen, channelTypeEncoder,
+			channelTypeDecoder))
+	}
+	tlv.SortRecords(tlvRecords)
+
+	tlvStream, err := tlv.NewStream(tlvRecords...)
+	if err != nil {
+		return err
+	}
+
+	return tlvStream.Encode(w)
+}
+
+// Decode deserializes the serialized DynPropose stored in the passed io.Reader
+// into the target DynPropose using the deserialization rules defined by the
+// passed protocol version.
+//
+// This is a part of the lnwire.Message interface.
+func (dp *DynPropose) Decode(r io.Reader, _ uint32) error {
+	// Parse out the only required field.
+	if err := ReadElements(r, &dp.ChanID); err != nil {
+		return err
+	}
+
+	// Parse out TLV stream.
+	var tlvRecords ExtraOpaqueData
+	if err := ReadElements(r, &tlvRecords); err != nil {
+		return err
+	}
+
+	// Prepare receiving buffers to be filled by TLV extraction.
+	var dustLimitScratch uint64
+	dustLimit := tlv.MakePrimitiveRecord(DPDustLimitSatoshis,
+		&dustLimitScratch)
+
+	var maxValueScratch uint64
+	maxValue := tlv.MakePrimitiveRecord(DPMaxHtlcValueInFlightMsat,
+		&maxValueScratch)
+
+	var reserveScratch uint64
+	reserve := tlv.MakePrimitiveRecord(DPChannelReserveSatoshis,
+		&reserveScratch)
+
+	var htlcMinScratch uint64
+	htlcMin := tlv.MakePrimitiveRecord(DPHtlcMinimumMsat, &htlcMinScratch)
+
+	var csvDelayScratch uint16
+	csvDelay := tlv.MakePrimitiveRecord(DPToSelfDelay, &csvDelayScratch)
+
+	var maxHtlcsScratch uint16
+	maxHtlcs := tlv.MakePrimitiveRecord(DPMaxAcceptedHtlcs,
+		&maxHtlcsScratch)
+
+	var fundingKeyScratch btcec.PublicKey
+	fundingKey := tlv.MakePrimitiveRecord(DPFundingPubkey,
+		&fundingKeyScratch)
+
+	var chanTypeScratch ChannelType
+	chanType := tlv.MakeDynamicRecord(DPChannelType, &chanTypeScratch,
+		dp.ChannelType.featureBitLen, channelTypeEncoder,
+		channelTypeDecoder)
+
+	// Create set of Records to read TLV bytestream into.
+	records := []tlv.Record{dustLimit, maxValue, reserve, htlcMin, csvDelay,
+		maxHtlcs, fundingKey, chanType}
+	tlv.SortRecords(records)
+
+	// Read TLV stream into record set.
+	extraBytesReader := bytes.NewReader(tlvRecords)
+	tlvStream, err := tlv.NewStream(records...)
+	if err != nil {
+		return err
+	}
+	typeMap, err := tlvStream.DecodeWithParsedTypesP2P(extraBytesReader)
```

Comment:
can use `ExtractRecords`?

Reply:
The issue with ExtractRecords is that I can't use it without RecordProducer instances for everything. I thought about it but it does require me to newtype all of the fields and define RecordProducer instances of them. I can do that, but I opted for a much more direct approach here. I'm not opposed to newtyping, generally I think it's a good practice, however there is something to be said here which is that the RecordProducer interface assumes that the TLV type number is the the same for every usage of that type regardless of which message it appears in which deviates from the semantics of the spec, which gave me additional hesitation on using it.